### PR TITLE
Dashboard: Add max time back and max time span settings

### LINKED ISSUE #4288: hard limits on time range ( controlled by operator?)
currently there's a quick range of 5years back, but using the calendar picker and zoom out controls, you can go arbitrarily far back (e.g. 30 or 100 years even)

however:
* the tsdb may only have data up to a certain retention (a year, 3 months, ... ?) so it doesn't make sense to go back further
* performance could degrade significantly or even tsdb crash if you go too far back.

so i propose admin controls to:
1) limit how far back one can go
2) what the max time range is

this would result in disabling of certain quick range options, restricting how far "zoom out" can go, and shielding users from certain timeranges in the calendar picker/editor.

we could perhaps tie this in to the quota config system because available retentions can be dependent on which org you're in.

### PR DESCRIPTION
**Pull request by ahlaw**
**What this PR does / why we need it**:

Currently, the time period for a dashboard can be modified via the time picker drop-down and the zoom out controls, and there is no way to prevent the users from selecting arbitrarily far and large time ranges. This PR adds dashboard settings that limits the time range selected via the time range controls.

Two new options will be available to admins under Time Options in the Grafana Dashboard Settings: **max time back** and **max time span**. The input for max time back is a time span that is the limit of how far back in time a selected time range can cover, relative to now. The input for max time span is a time span that represents the maximum length of time a selected time range can span. The input for both of these options must be valid time spans or empty, otherwise the option will default to the last valid input. By default, both inputs are empty. If either input is empty, there will be no restriction for their corresponding setting.

image-2

From a user perspective, there will be no change in the time range controls UI. If the user picks or zooms out to a valid time range, then the dashboard should update exactly as before. If the user picks an invalid time range, then the dashboard should not update, the current time range settings should remain the same, and an error message should appear in the time control dropdown instead of the dropdown closing. If the user clicks the zoom out control, it should only zoom out if the new time range would be valid, otherwise no action is taken. If the user enters an invalid time range via URL time control, they will be taken to the default time range. If the user inputs a setting that makes their current time range invalid, the dashboard will not update, but attempting to apply the current time range again will follow the invalid path. Note that the default time range of 6 hours before now will always be reachable via a page load.

image-3

Note that these settings only affect the time range controls and can be overridden by individual panel settings and URL time control query parameters.

**Which issue(s) this PR fixes**:

Resolves #4288 

cc: @alolita

### CODE REVIEW

**Review by hugohaggmark — COMMENTED** [review]
Thak you @ahlaw for contributing to this. I think it needs a bit more work and I'm not super familiar with these parts of the code base so I'm going to add @mckn as a reviewer too.

**Comment by hugohaggmark** [comment]
How to solve this for Explore, though?

File: public/app/features/dashboard/components/DashNav/DashNavTimeControls.tsx [hunk 1]
```diff
@@ -114,7 +129,10 @@ class UnthemedDashNavTimeControls extends Component<Props> {
           onMoveBackward={this.onMoveBack}
           onMoveForward={this.onMoveForward}
           onZoom={this.onZoom}
+          onClose={this.onClose}
           onChangeTimeZone={this.onChangeTimeZone}
+          invalid={this.state.isError}
+          error={'Invalid time range'}
```
Comment (dprokop) [inline public/app/features/dashboard/components/DashNav/DashNavTimeControls.tsx:135]: I would like to see some more details about this error here, to guide user what is the configured allowed range

File: public/app/features/dashboard/services/TimeSrv.ts [hunk 2]
```diff
@@ -303,6 +319,44 @@ export class TimeSrv {
...
+  private isValidTimeSpan(time: RawTimeRange): boolean {
+    const { maxTimeSpan } = this.dashboard?.timepicker ?? { maxTimeSpan: undefined };
+    const timeRange = rangeUtil.convertRawToRange(time);
+    const timeSpan = timeRange.to.valueOf() - timeRange.from.valueOf();
+    // Add 1 to maxTimeBackSpan to allow for rounding errors
+    const exceededMaxTimeSpan = maxTimeSpan && timeSpan > rangeUtil.intervalToMs(maxTimeSpan) + 1;
+    return !exceededMaxTimeSpan;
+  }
+
+  isValidTime(time: RawTimeRange) {
```
Comment (dprokop) [inline public/app/features/dashboard/services/TimeSrv.ts:357]:
```suggestion
  isValidTimeRange(time: RawTimeRange) {
```

**Comment by dprokop** [comment]
> How to solve this for Explore, though? 

Good point. Don't think we need to solve this for Explore in this PR, but worth starting a conversation. Would it be an idea to enable data source to control the allowed time ranges? It does not make much sense in the dashboard context as there can be multiple ds having independent, allowed ranges, and this would cause some weird logic to decide what's dashboard allowed range.

Questions I don't have an answer to is if data sources allow getting the retention info easily.

Looping in @davkal and @aocenas to see if this feature is something the O11y squad has on their agenda.

**Comment by aocenas** [comment]
@dprokop From what I seen in the linked issue I wonder if we even need this in explore. From what I understand the issue is more about having a dashboard which either won't show anything if you go out of some range or can actually bring down the DB but I assume that mainly happens in bigger dashboards and should not happen with one spurious query.

Also my assumption is that dashboards are setup by a power user for less savvy users so baking some limits makes more sense. In explore you usually already need to be more experienced with the datasource to be able to create query. Also from a stand point of permissions read only users can still change current time range of a dashboard right? Can they query through explore?

**Comment by oddlittlebird** [comment]
> Can they query through explore?

Viewers cannot access explore unless viewers_can_edit is enabled.

File: public/app/features/dashboard/components/DashboardSettings/SettingsCtrl.ts [hunk 3]
```diff
@@ -263,11 +263,21 @@ export class SettingsCtrl {
     this.renderCount++;
   };
 
+  onMaxTimeSpanChange = (maxTimeSpan: string) => {
+    this.dashboard.timepicker.maxTimeSpan = maxTimeSpan;
+    this.renderCount++;
+  };
+
```
[the remark this answers is not in the document]
Comment (ahlaw) [reply public/app/features/dashboard/components/DashboardSettings/SettingsCtrl.ts:270]: I'm not sure I understand this. AutoRefreshIntervals updates the dashboard model using the following handler without validation
https://github.com/grafana/grafana/blob/0ee424072863584fa8ea5ea5376a0726ad6f1678/public/app/features/dashboard/components/DashboardSettings/SettingsCtrl.ts#L256
The validation for maxTimeSpan and AutoRefreshIntervals are more or less in the same place, except the fact that AutoRefreshIntervals packages its field and logic in its own React component. Also, nowDelay validates in the same place. Unless you're suggesting these fields to be in their own separate component as well?

File: public/app/features/dashboard/components/DashNav/DashNavTimeControls.tsx (hunk 1 above)
Comment (ahlaw) [reply public/app/features/dashboard/components/DashNav/DashNavTimeControls.tsx:135]: Good idea, how does displaying the setting itself sound (e.g. Invalid time range: The maximum time back is 5h from now)? Also, would it make more sense to display the first violated rule or simply inform the user of both time limits?

File: public/app/features/dashboard/components/DashboardSettings/SettingsCtrl.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (hugohaggmark) [reply public/app/features/dashboard/components/DashboardSettings/SettingsCtrl.ts:270]: I'm suggesting that we need some function that validates the input from the user as a valid value. If you want to wrap that logic into a separate component that would be great but not what I'm suggesting.

File: docs/sources/dashboards/time-range-controls.md [hunk 4]
```diff
@@ -103,6 +103,8 @@ To access the dashboard time settings, click the **Dashboard settings** (gear) i
   - Standard [ISO 8601 time zones](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones), including UTC.
 - **Auto-refresh -** Customize the options displayed for relative time and the auto-refresh options. Entries are comma separated and accept any valid time unit.
 - **Now delay now- -** Override the `now` value by entering a time delay. Most commonly, this feature is used to accommodate known delays in data aggregation to avoid null values.
+- **Max time back -** Prevents the users from setting the dashboard `from time` to a value past the max time back. By default, the max time back is not set. The input should be the relative time back value, followed by a unit suffix (ms, s, m, h, d, w, M, y), e.g. `1y`. This will be interpreted as `now-1y` by Grafana.
```
Comment (hugohaggmark) [inline docs/sources/dashboards/time-range-controls.md:106]: `Max time back` doesn't seem that user friendly, WDTH @oddlittlebird?

Comment (oddlittlebird) [reply docs/sources/dashboards/time-range-controls.md:106]: Agreed. I thought I was getting time back at first glance. 

Maybe something like "Limit time range"? We need to make clear that we are limiting the ability of users to query more than XX amount of time in the past, or restricting the value they can inter in the From field.

Comment (ahlaw) [reply docs/sources/dashboards/time-range-controls.md:106]: I think "Time [range] lower limit" would be the most accurate description, but I also think using either the word "past" or "back" makes it more clear to the user how it is calculated/what kind of input is expected (i.e. how far back from now a user can go). Any other suggestions?

**Comment by torkelo** [comment]
Not sure about this for a couple of reasons. 

1) Where should this setting live? Per dashboard, per data source, per org or server instance? With the current approach, this will not be configurable / have any effect on explore or dashboard panels that have a panel relative time override (maybe less of a problem as you need edit permissions for this) 
2) Should this be enforced on the backend?

**Comment by ahlaw** [comment]
> Not sure about this for a couple of reasons.
> 
> 1. Where should this setting live? Per dashboard, per data source, per org or server instance? With the current approach, this will not be configurable / have any effect on explore or dashboard panels that have a panel relative time override (maybe less of a problem as you need edit permissions for this)
> 2. Should this be enforced on the backend?

1. I believe per dashboard is the preferred setting level because the primary use case was to provide safeguards for less savvy users using public dashboards. Panel relative time override should be fine because edit permissions implies they know what they're doing.

2. I'm not too familiar with the code base, but my reasoning was that we needed to modify behaviour of specific user controls based on the dashboard model, and the controls logic is in the frontend. Would definitely love more input on this though.

File: docs/sources/dashboards/time-range-controls.md (hunk 4 above)
Comment (achatterjee-grafana) [reply docs/sources/dashboards/time-range-controls.md:106]: "Time [range] lower limit" is also not clear to me. 

This is also not clear: "Prevents the users from setting the dashboard `from time` to a value past the max time back." What does `from time` mean?

I will try to offer a better wording suggestion.

File: docs/sources/dashboards/time-range-controls.md [hunk 5]
```diff
@@ -103,6 +103,8 @@ To access the dashboard time settings, click the **Dashboard settings** (gear) i
...
+- **Max time span -** Prevents the users from setting the dashboard time range to an interval greater than the max time span. By default, the max time span is not set. The input should be the max span value, followed by a unit suffix (ms, s, m, h, d, w, M, y), e.g. `5y`.
```
Comment (achatterjee-grafana) [inline docs/sources/dashboards/time-range-controls.md:107]: Change to: Prevents the user from setting the dashboard time range to an interval that is greater than the max time span.

**Review by achatterjee-grafana — COMMENTED** [review]
Added a few copy-edit suggestions.

File: docs/sources/dashboards/time-range-controls.md (hunk 4 above)
Comment (ahlaw) [reply docs/sources/dashboards/time-range-controls.md:106]: Lower limit and `from time` are both attempts to reference the start of the time range, and [range] meant to suggest that the word range could be included or excluded in that phrase in actual use to clear any confusion. How about "time range start" and "time range end" for all references to the endpoints of the time range?

**Comment by achatterjee-grafana** [comment]
Can someone explain what "time back" means in the context of the "Max time back" property? Thanks.

File: docs/sources/dashboards/time-range-controls.md (hunk 4 above)
Comment (achatterjee-grafana) [reply docs/sources/dashboards/time-range-controls.md:106]: I like this idea. Thank you for the clarification.

Comment (achatterjee-grafana) [reply docs/sources/dashboards/time-range-controls.md:106]: On a separate note, can we revise the "**Now delay now- -**" phrase? Maybe "Add delay to now"?

Comment (ahlaw) [reply docs/sources/dashboards/time-range-controls.md:106]: Maybe out of scope for this PR, it matches the setting name and should be a noun. I think it is trying to do the same thing, which is to clarify that the setting is relative to 'now'. As for the original concern with the setting name, does "Time range start limit" seem like a reasonable alternative to "Max time back"? And then in the tooltip/docs clarify that the actual point in time is calculated as "now - \<setting>"?

Comment (achatterjee-grafana) [reply docs/sources/dashboards/time-range-controls.md:106]: "Time range start limit" makes sense, let us go ahead with it.

**Review by achatterjee-grafana — APPROVED** [review]
LGTM

**Review by hugohaggmark — CHANGES_REQUESTED** [review]
@ahlaw thank you for this PR, we've had some discussions and we need some time to think about the best architecture to solve this issue.

**Comment by ahlaw** [comment]
Thanks for the review effort. Any specific reason why this was closed?

**Pull request closed without merging**